### PR TITLE
Allow creating a list of lists from an array of arrays

### LINKED ISSUE #24277: Feature request: list initialisation
Initialisation of a list using this syntax does not compile.

var lst3 : list(list(string)) = [["this", "that", "the other"],["one","two","three"]];

### Summary of Problem
<!--
What behavior did you observe when encountering this issue?

$CHPL_HOME/modules/standard/List.chpl:519: In method '_commonInitFromIterable':
$CHPL_HOME/modules/standard/List.chpl:522: error: unresolved call 'list(list(string,false),false).pushBack([domain(1,int(64),one)] string)'
$CHPL_HOME/modules/standard/List.chpl:890: note: this candidate did not match: ref list.pushBack(other: [?d] eltType)
$CHPL_HOME/modules/standard/List.chpl:890: note: because an argument was incompatible
$CHPL_HOME/modules/standard/List.chpl:522: note: other candidates are:
$CHPL_HOME/modules/standard/List.chpl:820: note:   ref list.pushBack(pragma"no auto destroy"in x: this.eltType)
$CHPL_HOME/modules/standard/List.chpl:868: note:   ref list.pushBack(other: list(eltType, ?p))
$CHPL_HOME/modules/standard/List.chpl:914: note:   ref list.pushBack(other: range(eltType, ?b, ?d))
  $CHPL_HOME/modules/standard/List.chpl:455: called as (list(list(string,false),false))._commonInitFromIterable(iterable: [domain(1,int(64),one)] [domain(1,int(64),one)] string) from method 'init='
  issu.chpl:2: called as (list(list(string,false),false)).init=(other: [domain(1,int(64),one)] [domain(1,int(64),one)] string)
note: generic instantiations are underlined in the above callstack 

What behavior did you expect to observe?
Successful compilation.

Is this a blocking issue with no known work-arounds?
No.

### Steps to Reproduce

**Source Code:**
list_1.chpl

```chapel
use List;

var lst3 : list(list(string)) = [["this", "that", "the other"],["one","two","three"]];

```

**Compile command:**
chpl list_1.chpl

**Execution command:**
Does not compile.

### Configuration Information

- Output of `chpl --version`:
- chpl version 1.33.0
  built with LLVM version 15.0.7
  available LLVM targets: xcore, x86-64, x86, wasm64, wasm32, ve, systemz, sparcel, sparcv9, sparc, riscv64, riscv32, ppc64le, ppc64, ppc32le, ppc32, nvptx64, nvptx, msp430, mips64el, mips64, mipsel, mips, lanai, hexagon, bpfeb, bpfel, bpf, avr, thumbeb, thumb, armeb, arm, amdgcn, r600, aarch64_32, aarch64_be, aarch64, arm64_32, arm64
Copyright 2020-2023 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)

- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
CHPL_TARGET_PLATFORM: freebsd
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: amd64
CHPL_TARGET_CPU: unknown
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: fifo
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: bundled
CHPL_HWLOC: none
CHPL_RE2: bundled
CHPL_LLVM: system
CHPL_AUX_FILESYS: none

- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
- FreeBSD clang version 14.0.5 (https://github.com/llvm/llvm-project.git llvmorg-14.0.5-0-gc12386ae247c)
Target: x86_64-unknown-freebsd13.2
Thread model: posix
InstalledDir: /usr/bin

- (For Cray systems only) Output of `module list`:

### PR DESCRIPTION
**Pull request by jabraham17**
This PR allows creating a `list` of `list`s from an array of arrays. Previosuly, the following resulted in a confusing error to users.

```chapel
var a: list(list(string)) = [["a", "b"], ["c", "d"]];
```

To resolve the issue, users needed to write the following to workaround this.

```chapel
var a: list(list(string)) = [new list(["a", "b"]), new list(["c", "d"])];
```

This PR extends the existing code that allows implicit list creation from arrays to handled arrays and arrays, so that both of the above code snippets work the same. This is also extendable to arrays of arrays of arrays.

Testing
- [x] paratest without comm
- [x] paratest with comm

Resolves #24277

[Reviewed by @lydia-duncan]

### CODE REVIEW

**Review by lydia-duncan — APPROVED** [review]
Nice!  We just want to make sure that `list(array(something))` doesn't run into problems as a result of this change.

**Comment by jabraham17** [comment]
Looks like it does

```chapel
use List;
var a = new list([[1,2,3],[4,5,6]]);
compilerWarning((a.type:string));
writeln(a);
```

This code works on both main and this branch


This compiles on both main and this branch, but segfaults at runtime for both main and this branch. Not sure whats going on there...

```chapel
use List;
var a: list([0..<3] int) = [[1,2,3],[4,5,6]];
compilerWarning((a.type:string));
writeln(a);
```


I will add both of these as tests before I merge this PR (the latter being a future)

File: test/library/standard/List/initEquals/listInitEqualsArrayBug.bad
Comment (lydia-duncan) [inline test/library/standard/List/initEquals/listInitEqualsArrayBug.bad:3]: This might have different behavior on Macs versus Linux, but I think it's worth trying to have a .bad file and seeing what happens

Comment (jabraham17) [reply test/library/standard/List/initEquals/listInitEqualsArrayBug.bad:3]: This passed as a future on my Mac and timed-out on Linux. I can just remove the `.bad` file to prevent it failing in nightly

Comment (lydia-duncan) [reply test/library/standard/List/initEquals/listInitEqualsArrayBug.bad:3]: Sounds good, I think this is good to go once that adjustment happens

File: test/library/standard/List/initEquals/listInitEqualsArrayBug.future
Comment (cassella) [inline test/library/standard/List/initEquals/listInitEqualsArrayBug.future:1]: Is there a github issue for this?

Comment (jabraham17) [reply test/library/standard/List/initEquals/listInitEqualsArrayBug.future:1]: This bug is captured in #24305

**Pull request merged**
